ecc/bls12-377/signature/schnorr/schnorr.go

```diff
@@ -0,0 +1,286 @@
+		hash = hash[:sizeFr]
+	}
+	ret := new(big.Int).SetBytes(hash)
+	excess := len(hash)*8 - sizeFr
+	if excess > 0 {
+		ret.Rsh(ret, uint(excess))
+	}
+	return ret
+}
+
+type zr struct{}
+
+// Read replaces the contents of dst with zeros. It is safe for concurrent use.
+func (zr) Read(dst []byte) (n int, err error) {
+	for i := range dst {
+		dst[i] = 0
+	}
+	return len(dst), nil
+}
+
+var zeroReader = zr{}
+
+const (
+	aesIV = "gnark-crypto IV." // must be 16 chars (equal block size)
+)
+
+func nonce(privateKey *PrivateKey, hash []byte) (csprng *cipher.StreamReader, err error) {
+	// This implementation derives the nonce from an AES-CTR CSPRNG keyed by:
+	//
+	//    SHA2-512(privateKey.scalar ∥ entropy ∥ hash)[:32]
+	//
+	// The CSPRNG key is indifferentiable from a random oracle as shown in
+	// [Coron], the AES-CTR stream is indifferentiable from a random oracle
+	// under standard cryptographic assumptions (see [Larsson] for examples).
+	//
+	// [Coron]: https://cs.nyu.edu/~dodis/ps/merkle.pdf
+	// [Larsson]: https://web.archive.org/web/20040719170906/https://www.nada.kth.se/kurser/kth/2D1441/semteo03/lecturenotes/assump.pdf
+
+	// Get 256 bits of entropy from rand.
+	entropy := make([]byte, 32)
+	_, err = io.ReadFull(rand.Reader, entropy)
+	if err != nil {
+		return
+
+	}
+
+	// Initialize an SHA-512 hash context; digest...
+	md := sha512.New()
+	md.Write(privateKey.scalar[:sizeFr]) // the private key,
+	md.Write(entropy)                    // the entropy,
+	md.Write(hash)                       // and the input hash;
+	key := md.Sum(nil)[:32]              // and compute ChopMD-256(SHA-512),
+	// which is an indifferentiable MAC.
+
+	// Create an AES-CTR instance to use as a CSPRNG.
+	block, _ := aes.NewCipher(key)
+
+	// Create a CSPRNG that xors a stream of zeros with
+	// the output of the AES-CTR instance.
+	csprng = &cipher.StreamReader{
+		R: zeroReader,
+		S: cipher.NewCTR(block, []byte(aesIV)),
+	}
+
+	return csprng, err
+}
+
+// Equal compares 2 public keys
+func (pub *PublicKey) Equal(x signature.PublicKey) bool {
+	xx, ok := x.(*PublicKey)
+	if !ok {
+		return false
+	}
+	bpk := pub.Bytes()
+	bxx := xx.Bytes()
+	return subtle.ConstantTimeCompare(bpk, bxx) == 1
+}
+
+// Public returns the public key associated to the private key.
+func (privKey *PrivateKey) Public() signature.PublicKey {
+	var pub PublicKey
+	pub.A.Set(&privKey.PublicKey.A)
+	return &pub
+}
+
+// Sign performs the Schnorr signature
+//
+// k ← 𝔽r (random)
+// P = k ⋅ g1Gen
+// r = H ( P || m )
+// s = k - x ⋅ r
+// signature = {r, s}
+func (privKey *PrivateKey) Sign(message []byte, hFunc hash.Hash) ([]byte, error) {
+	scalar, r, s := new(big.Int), new(big.Int), new(big.Int)
+	scalar.SetBytes(privKey.scalar[:sizeFr])
+	csprng, err := nonce(privKey, message)
+	if err != nil {
+		return nil, err
+	}
+	k, err := randFieldElement(csprng)
+	if err != nil {
+		return nil, err
+	}
+
+	var P bls12377.G1Affine
+	P.ScalarMultiplicationBase(k)
+
+	if hFunc != nil {
```

Comment:
Hmm, I think this is incorrect. In Schnorr we use the hash for Fiat-Shamir challenge. If `hFunc==nil`, then we do not hash signer randomness into challenge and get a weak instance. Depending on the attack, we either leak secret key or allow forging signatures.

In description, we should rather think as the challenge as `H1(P || H2(m))`, where `hFunc` refers to `H2` not `H1`. `H1` should be fixed per-curve to fill the whole scalar element.

Reply:
I thought about it. I think the secret key is safe, but still allows forgeability.

Reply:
I think there is no point into hashing the message (H2) before hashing its concatenation with P (H1), as we consider arbitrary long messages. My initial thought was that if the user uses `nil` as `hFunc` it would mean he did `H(P || m)` prior to signing, and we just convert it to and integer. But I agree that if it opens room to ambiguity we should clear this up. So either:
- We remove the `hFunc` argument from the interface and enforce a hash function inside `Sign()` and `Verify()` for all signature schemes --> but we wouldn't be able to replace with a SNARK-friendly hash in gnark if needed, or
- if `hFunc == nil` then we use an encoded hash inside `Sign()` and `Verify()`, otherwise we use the `hFunc` provided by the caller --> but we would always compute a hash and can't delegate this (e.g. out-circuit in SNARKs), or
- `hFunc` would mean H1 as in your example and we always do H2 --> same we can't delegate H2 but the interface and implementation would be similar across signature schemes.

Reply:
One option is if there is some nice SNARK-friendly instantiation of the Fiat-Shamir challenger for Schnorr signatures. https://eprint.iacr.org/2020/915.pdf says there is (Theorem 2.4 and Construction 2.5), but as I understand, imo requires a (one-time?) trusted setup. But if it is true, then we can fix `H1` and `hFunc` gives `H2`.